fcrepo-clients/src/main/java/edu/unc/lib/dl/util/DerivativeService.java

```diff
@@ -0,0 +1,156 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package edu.unc.lib.dl.util;
+
+import static edu.unc.lib.dl.fcrepo4.RepositoryPathConstants.HASHED_PATH_DEPTH;
+import static edu.unc.lib.dl.fcrepo4.RepositoryPathConstants.HASHED_PATH_SIZE;
+import static edu.unc.lib.dl.fcrepo4.RepositoryPaths.idToPath;
+import static edu.unc.lib.dl.model.DatastreamType.getByIdentifier;
+import static edu.unc.lib.dl.model.StoragePolicy.EXTERNAL;
+import static org.springframework.util.Assert.notNull;
+
+import java.io.File;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import edu.unc.lib.dl.fedora.PID;
+import edu.unc.lib.dl.model.DatastreamType;
+
+/**
+ * Service which locates and returns file information about derivatives for
+ * repository objects.
+ *
+ * @author bbpennel
+ *
+ */
+public class DerivativeService {
+
+    private String derivativeDir;
+
+    private volatile static List<DatastreamType> derivativeTypes;
+
+    public DerivativeService() {
+    }
+
+    /**
+     * Gets the derivative of type dsType for the object identified by pid.
+     *
+     * @param pid pid of the object. Required.
+     * @param dsType type of the derivative to retrieve.
+     * @return Derivative of type dsType for object pid, or null if the
+     *         derivative does not exist.
+     */
+    public Derivative getDerivative(PID pid, DatastreamType dsType) {
+        notNull(pid);
+        notNull(dsType);
+
+        String id = pid.getId();
+        String hashedPath = idToPath(id, HASHED_PATH_DEPTH, HASHED_PATH_SIZE);
+
+        // Construct the full path of the derivative
+        Path derivPath = Paths.get(derivativeDir, dsType.getId(), hashedPath, id + "." + dsType.getExtension());
+        File derivFile = derivPath.toFile();
+
+        // If the derivative file does not exist, then return no result
+        if (!derivFile.exists()) {
+            return null;
+        }
+
+        return new Derivative(dsType, derivFile);
+    }
+
+    /**
+     * Generates a list of all derivatives present for the object identified by
+     * pid.
+     *
+     * @param pid pid of the object. Required.
+     * @return list of derivatives for pid
+     */
+    public List<Derivative> getDerivatives(PID pid) {
+        notNull(pid);
+
+        return listDerivativeTypes().stream()
+            .map(derivType -> getDerivative(pid, derivType))
+            .filter(deriv -> deriv != null)
+            .collect(Collectors.toList());
+    }
+
+    /**
+     * List all DatastreamTypes which are considered derivatives.
+     *
+     * @return
+     */
+    public static List<DatastreamType> listDerivativeTypes() {
+        if (derivativeTypes == null) {
+            derivativeTypes = Arrays.stream(DatastreamType.values())
+                    .filter(dt -> EXTERNAL.equals(dt.getStoragePolicy()))
```

Comment:
so all externally-stored datastream types are derivative types?

Reply:
For the time being, the criteria might change later, which is why I wanted it abstracted into a helper method like this.